src/Renci.SshNet/PrivateKeyFile.cs

```diff
@@ -717,6 +737,265 @@ private static Key ParseOpenSshV1Key(byte[] keyFileData, string passPhrase)
+        /// <param name="keyData">The key data.</param>
+        /// <remarks>
+        /// The format of PKCS#8 DSA (and other) private keys is not well documented:
+        /// it is hidden away in PKCS#11 v2.01, section 11.9.
+        /// OpenSSL's default DSA PKCS#8 private key format complies with this standard.
+        /// <see href="https://docs.openssl.org/1.1.1/man1/pkcs8/#standards" />
+        /// <see href="https://www.cryptsoft.com/pkcs11doc/STANDARD/v201-95.pdf" />.
+        /// <code>
+        /// Dss-Parms ::= SEQUENCE {
+        ///     p INTEGER,
+        ///     q INTEGER,
+        ///     g INTEGER
+        /// }
+        /// </code>
+        /// DSAprivatekeys are represented as BER-encoded ASN.1 type INTEGER.
+        /// </remarks>
+        /// <returns>The <see cref="DsaKey" />.</returns>
+        private static DsaKey ParseDSAPrivateKey_Pkcs8_OpenSSL(byte[] parametersData, byte[] keyData)
+        {
+            var parametersReader = new AsnReader(parametersData, AsnEncodingRules.BER);
+            var sequenceReader = parametersReader.ReadSequence();
+            parametersReader.ThrowIfNotEmpty();
+
+            var p = sequenceReader.ReadInteger();
+            var q = sequenceReader.ReadInteger();
+            var g = sequenceReader.ReadInteger();
+            sequenceReader.ThrowIfNotEmpty();
+
+            var keyReader = new AsnReader(keyData, AsnEncodingRules.BER);
+            var x = keyReader.ReadInteger();
+            keyReader.ThrowIfNotEmpty();
+
+            var y = BigInteger.ModPow(g, x, p);
+
+            return new DsaKey(p, q, g, y, x);
+        }
+
+        /// <summary>
+        /// Parses OpenSSL EC private key.
+        /// </summary>
+        /// <param name="keyData">The key data.</param>
+        /// <remarks>
+        /// OpenSSL produces ASN.1 DER encoded SEC1 private key.
+        /// <see href="https://docs.openssl.org/1.1.1/man1/ec/#description" />
+        /// <see href="https://www.secg.org/sec1-v2.pdf" />.
+        /// Republished as RFC5915 <see href="https://www.rfc-editor.org/rfc/rfc5915" />,
+        /// See answer at <see href="https://crypto.stackexchange.com/a/48709" />.
+        /// <code>
+        /// ECPrivateKey ::= SEQUENCE {
+        ///     version INTEGER { ecPrivkeyVer1(1) } (ecPrivkeyVer1),
+        ///     privateKey OCTET STRING,
+        ///     parameters [0] ECDomainParameters {{ SECGCurveNames }} OPTIONAL,
+        ///     publicKey [1] BIT STRING OPTIONAL
+        /// }
+        /// </code>
+        /// When generating a transfer encoding, generators SHOULD use
+        /// Distinguished Encoding Rules (DER) and receivers SHOULD be
+        /// prepared to handle Basic Encoding Rules (BER) and DER.
+        /// </remarks>
+        /// <returns>The <see cref="EcdsaKey" />.</returns>
+        private static EcdsaKey ParseECPrivateKey_SEC1(byte[] keyData)
+        {
+            var keyReader = new AsnReader(keyData, AsnEncodingRules.DER);
```

Comment:
Based on the comment above, should this be

```suggestion
            var keyReader = new AsnReader(keyData, AsnEncodingRules.BER);
```

?

Reply:
DERs are subset of BERs. I think DERs are enough here since the data is generated using DERs?

Reply:
I am reading this part of the comment: "receivers SHOULD be prepared to handle Basic Encoding Rules (BER) and DER"

In my understanding, that means using AsnEncodingRules.BER to match the comment

Reply:
Actually I'm confused about the doc. Do you know why it says "BER **and** DER".

Reply:
I guess they are trying to be explicit but they do not make it any clearer by doing so